components/epaxos/src/protocol/replica.rs

```diff
@@ -0,0 +1,157 @@
+
+        // TODO(lsl): JoinHandle may not enough, we need a wrapper to contain JoinHandle and data used by
+        // exec thread
+        exec_worker: JoinHandle, // handle of exec thread
+        maxSeq: int32, // max seq ever known in cluster
+
+        // FIXME(lsl): what is it used for ? see only assignment but never used
+        // latestCPReplica: ReplicaID, // record the replica peer in the latest communication
+        // latestCPInstance: InstanceNum,    // record the instance id in the latest communication
+        // combine the above two fields in latestCP
+        latestCP: InstanceID, // record the instance id in the lastest communication
+
+        // FIXME(lsl): this channel is used to do recovery.
+        // as discussed, we use a in ballot number(epoch.a.replica) as a hint of recovery,
+        // so instancesToRecover is not necessary to us.
+        // instancesToRecover: chan *instanceId,
+}
+
+impl Replica {
+    /// create a new Replica
+    /// do all the initialization and start all necessary threads here,
+    /// so after this call, replica is fully functional.
+    pub fn new(replica: ReplicaID, cluster: &ClusterInfo, thrifty: bool, exec: bool, beacon: bool,) -> Result<Replica, Error> {
+    }
+
+    /// handle a client request
+    /// this is the main logic to implement choosing commands in epaxos protocol
+    /// as described in figure 2.
+    pub fn propose(&mut self, cmds: &Vec<Command>) {
+    }
+
+    /// do graceful shutdown of epaxos replica part
+    pub fn shutdown() {
+    }
+
+    /// send a message
+    /// if conn is None, make connection
+    /// TODO(lsl): we really need something like a request context
+    ///            to store conn and anything should be shared in a request or proposing an instance
+    fn send_msg(&mut self, conn: Option(&mut TcpStream), to: ReplicaID, msg: &Message) -> (&mut TcpStream, Result<Message, Error>) {
+    }
+
+    /// start exec thread
+    fn start_exec_thread(&mut self) {
+    }
+
+    /// this is the main logic to implement recovery procedure in epaxos protocol
+    /// as described in figure 3.
+    fn explicit_prepare(&mut self, instance: &Instance) {
+    }
+
+    // FIXME(lsl): these methods predestined to run in multi-thread and change self,
+    //             so is it good to implement as a method or a function to take Replica as arg?
+    //             but for now, it doesn't matter since we just want the interface.
+
+    /// send prepare request and receive reponse, return updated reponse or error.
+    fn make_prepare_req(&mut self, cmds: &Vec<Command>) -> Result<PrepareReply, Error> {
+        // ReplicaID and ballot are stored in self.
+    }
+
+    // ASK(lsl->xp): is it ok to return a result and ask http-framework to judge type and send
+    // reply to requester?
+    // I think that would be a great place to do statistics and measurement.
+    // and by making Replica independent,it's easier to do test.
```

Comment:
内部注释用 // 吧
接口文档注释用 /// 吧

Reply:
我看到你这个struct的函数，有的用///有的用//，不统一

Reply:
写在struct头上的是///，内部的注释用的是 //。
没有不统一

Reply:
这个struct有的函数用的///，有的是//，有的两个交叉用
```
    /// this is the main logic to implement recovery procedure in epaxos protocol
    /// as described in figure 3.
    fn explicit_prepare(&mut self, instance: &Instance) {
    }

    // FIXME(lsl): these methods predestined to run in multi-thread and change self,
    //             so is it good to implement as a method or a function to take Replica as arg?
    //             but for now, it doesn't matter since we just want the interface.

    /// send prepare request and receive reponse, return updated reponse or error.
    fn make_prepare_req(&mut self, cmds: &Vec<Command>) -> Result<PrepareReply, Error> {
        // ReplicaID and ballot are stored in self.
    }
```